crates/ruff/src/directives.rs

```diff
@@ -217,6 +219,154 @@ fn extract_isort_directives(lxr: &[LexResult], locator: &Locator) -> IsortDirect
     }
 }
 
+/// A comment that contains a [`TodoDirective`]
+pub struct TodoComment<'a> {
+    /// The comment's text
+    pub content: &'a str,
+    /// The directive found within the comment.
+    pub directive: TodoDirective,
+    /// The comment's actual [`TextRange`].
+    pub range: TextRange,
+    /// The comment range's position in [`Indexer`].comment_ranges()
+    pub range_index: usize,
+}
+
+impl<'a> TodoComment<'a> {
+    /// Attempt to transform a normal comment into a [`TodoComment`].
+    pub fn from_comment(content: &'a str, range: &TextRange, range_index: usize) -> Option<Self> {
+        TodoDirective::extract_directive(content, range).map(|directive| Self {
+            directive,
+            content,
+            range: *range,
+            range_index,
+        })
+    }
+
+    /// Determine the starting location for the [`TodoDirective`], relative to the comment's
+    /// starting offset
+    pub fn directive_offset(&self) -> TextSize {
+        self.directive.range.start() - self.range.start()
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct TodoDirective {
+    /// The directive's literal text in the source code file.
+    pub content: String,
```

Comment:
Storing the content of every directive is kind of expensive because we need to allocate, copy, and later de-allocate the content. It is also somewhat redundant because we already store the `range`, which allows us to retrieve the text from the `Locator`. 

There are two ways to avoid that: 
* Store a reference to the `Locator` on `TodoDirective` and implement a `content(&self) -> &str` function that uses `locator.slice(self.range)` to get the content. The downside of this is that we need to store an extra pointer for every `TodoDirective` and that it now adds lifetimes everywhere. 
* Create a `content(&self, locator: &Locator) -> &str` method that returns the content (same as above). The downside here is that `dbg!(&directive)` no longer contains the content, and that calling `directive.content(locator)` is a bit less ergonomic.

Reply:
> Gotcha. I'll implement the latter way, but just for my own curiosity - is there a reason we couldn't also store the content as a &'a str similar to TodoComment other than the additional lifetimes?

Lol yeah, that would work too and is the better choice than storing a reference to `Locator`. It adds an additional 8 bytes but that's fine.